addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifGuid.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2021 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.addon.reports.sarif;
+
+import java.util.Objects;
+import java.util.UUID;
+
+/**
+ * Represents a GUID for Sarif
+ * https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317438
+ */
+public class SarifGuid {
+
+    private String guid;
+
+    private SarifGuid() {
+        // force to use factory methods
+    }
+
+    public static SarifGuid createCweGuid(int cweId) {
+        return createByProvider("" + cweId, SarifToolData.INSTANCE.getCwe());
+    }
+
+    public static SarifGuid createForTaxa(String identifier, SarifTaxonomy taxonomy) {
+        return createByProvider(identifier, taxonomy);
+    }
+
+    public static SarifGuid createToolcomponentGUID(SarifTaxonomyDataProvider component) {
+        return createByProvider("<<tool-component>>", component);
+    }
+
+    private static SarifGuid createByIdentifier(String identifier) {
+        SarifGuid sarifGuid = new SarifGuid();
+        UUID nameBasedUUID = UUID.nameUUIDFromBytes(identifier.getBytes());
+        sarifGuid.guid = nameBasedUUID.toString();
+        return sarifGuid;
+    }
+
+    /**
+     * Creates a SARFI guid - by using data from provider (name, taxonomy version) and the given id.
+     *
+     * @param id
+     * @param provider
+     * @return guid
+     */
+    private static SarifGuid createByProvider(String id, SarifTaxonomyDataProvider provider) {
+        // e.g. name:CWE:4.4:79
+        String identifier =
+                "name:" + provider.getName() + ":" + provider.getTaxonomyVersion() + ":" + id;
+        return createByIdentifier(identifier);
+    }
+
+    public String getGuid() {
+        return guid;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(guid);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null) return false;
+        if (getClass() != obj.getClass()) return false;
+        SarifGuid other = (SarifGuid) obj;
+        return Objects.equals(guid, other.guid);
+    }
```

Comment:
1. Braces: OK  
 This is eclipse generated code - I will use the template with "use blocks" now, so braces will be generated. Will change
2. "null check may as well be first."  
 Sorry, I dont' get this, after regenerating the code with braces in eclipse I got (remarks are from myself):
```java
	public boolean equals(Object obj) {
		if (this == obj) { // same instance
			return true;
		}
		if (obj == null) { // other object is null
			return false;
		}
                // no other null check necessary - because THIS object cannot be null...
		if (getClass() != obj.getClass()) {
			return false;
		}
		SarifGuid other = (SarifGuid) obj;
		return Objects.equals(guid, other.guid); // handles already guid other.guid being null
	}
```
So every null condition is handled correctly and there is no need to any additional operation.

Reply:
Not a huge deal with the null thing. I just figured `obj == null` being the first check might be ever so slightly more efficient, since if it's null them obviously aren't equal (same instance). But really it's like 6 of one half a dozen of another ;)